esp-hal/src/uart.rs

```diff
@@ -857,16 +886,23 @@ where
     }
 
     #[cfg(any(esp32c2, esp32c3, esp32s3))]
-    fn change_baud_internal(&self, baudrate: u32, clocks: &Clocks) {
-        // we force the clock source to be APB and don't use the decimal part of the
-        // divider
-        let clk = clocks.apb_clock.to_Hz();
+    fn change_baud_internal(&self, baudrate: u32, clock_source: ClockSource, clocks: &Clocks) {
+        let clk = match clock_source {
+            ClockSource::Apb => clocks.apb_clock.to_Hz(),
+            ClockSource::Xtal => clocks.xtal_clock.to_Hz(),
+            ClockSource::RcFast => RC_FAST_CLK.to_Hz(),
+        };
+
         let max_div = 0b1111_1111_1111 - 1;
         let clk_div = ((clk) + (max_div * baudrate) - 1) / (max_div * baudrate);
 
         T::register_block().clk_conf().write(|w| unsafe {
             w.sclk_sel()
-                .bits(1) // APB
+                .bits(match clock_source {
```

Comment:
probably selecting the source-clock should be done in `change_baud_internal` now
Reason is that the user could call `pub fn change_baud(...)` later and pass a different clock source. In that case the resulting baud-rate would be wrong

Reply:
just realized the web-diff UI confused me. sclk_sel is set in `change_baud_internal `  - sorry for the noise - all good